### PR TITLE
integrate with gcs for images uploading

### PR DESCRIPTION
**Pull request by nickhsine**
@hcchien

### CODE REVIEW

File: admin/server/api/gcs.js [hunk 1]
```diff
@@ -0,0 +1,69 @@
+var gcloud = require('gcloud');
+var keystone = require('../../../');
+var gcsHelper = require('../../../lib/gcsHelper');
```
Comment (hcchien) [inline admin/server/api/gcs.js]: should we have better variables sorting convention?

File: admin/server/api/gcs.js [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+  upload: function (req, res) {
+    if (req.files && req.files.file) {
+      var options = {};
+
+      var gcsConfig = keystone.get('gcs config');
+      var bucket = gcsHelper.initBucket(gcsConfig);
+
+      var destination = gcsConfig.destination || '';
+
+			var file = req.files.file;
```
Comment (hcchien) [inline admin/server/api/gcs.js]: wrong indent?

File: fields/types/gcsimage/GcsImageType.js [hunk 3]
```diff
@@ -0,0 +1,368 @@
...
+ */
+
+gcsimage.prototype.addToSchema = function () {
+
+	var field = this;
+	var schema = this.list.schema;
+
+	var paths = this.paths = {
+		// fields
+    bucket: this._path.append('.bucket'),
```
Comment (hcchien) [inline fields/types/gcsimage/GcsImageType.js]: wrong indent

File: fields/types/gcsimage/GcsImageType.js [hunk 4]
```diff
@@ -0,0 +1,368 @@
...
+		filename: this._path.append('.filename'),
+		originalname: this._path.append('.originalname'),
+		path: this._path.append('.path'),
+		size: this._path.append('.size'),
+		filetype: this._path.append('.filetype'),
+    url: this._path.append('.url'),
```
Comment (hcchien) [inline fields/types/gcsimage/GcsImageType.js]: wrong indent

File: fields/types/gcsimage/GcsImageType.js [hunk 5]
```diff
@@ -0,0 +1,368 @@
...
+	};
+
+	var schemaPaths = this._path.addTo({}, {
+    bucket: String,
+		filename: String,
+		originalname: String,
+	  path: String,
+		size: Number,
+		filetype: String,
+    url: String
```
Comment (hcchien) [inline fields/types/gcsimage/GcsImageType.js]: wrong indent

File: fields/types/gcsimage/GcsImageType.js [hunk 6]
```diff
@@ -0,0 +1,368 @@
...
+	};
+
+	// The .exists virtual indicates whether a file is stored
+	schema.virtual(paths.exists).get(function () {
+		return schemaMethods.exists.apply(this);
+	});
+
+	var reset = function (item) {
+		item.set(field.path, {
+      bucket: '',
```
Comment (hcchien) [inline fields/types/gcsimage/GcsImageType.js]: wrong indent

File: admin/server/api/gcs.js [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+
+            var destination = gcsConfig.destination || '';
+
+            var file = req.files.file;
+
+            // Upload a local file to a new file to be created in your bucket.
+            gcsHelper.uploadFileToBucket(bucket, file.path, {
+                destination: destination + file.originalname, // gcs.uploadDirectory
+            }).then(function(uploadedFile) {
+                return gcsHelper.makeFilePublicPrivateRead(uploadedFile, gcsConfig.publicRead);
```
Comment (hcchien) [inline admin/server/api/gcs.js]: Do you need to confirm if the gcsConfig.publicRead valid or not first?

File: fields/types/gcsimage/GcsImageColumn.js [hunk 8]
```diff
@@ -0,0 +1,15 @@
+var React = require('react');
+
+var GcsImageColumn = React.createClass({
+	render: function () {
+		var value = this.props.data.fields[this.props.col.path];
+		var isVal = value.url ? value.url : null;
```
Comment (hcchien) [inline fields/types/gcsimage/GcsImageColumn.js]: is it possible that the value is null or undefined?

File: fields/types/gcsimage/GcsImageType.js [hunk 9]
```diff
@@ -0,0 +1,361 @@
...
+        filename: this._path.append('.filename'),
+        originalname: this._path.append('.originalname'),
+        path: this._path.append('.path'),
+        size: this._path.append('.size'),
+        filetype: this._path.append('.filetype'),
+        url: this._path.append('.url'),
+        // virtuals
+        exists: this._path.append('.exists'),
+        upload: this._path.append('_upload'),
+        action: this._path.append('_action')
```
Comment (hcchien) [inline fields/types/gcsimage/GcsImageType.js]: Do you want to sort the keys? :p

File: fields/types/gcsimage/GcsImageType.js [hunk 10]
```diff
@@ -0,0 +1,361 @@
...
+
+    var reset = function(item) {
+        item.set(field.path, {
+            bucket: '',
+            filename: '',
+            originalname: '',
+            path: '',
+            size: 0,
+            filetype: '',
+            url: ''
```
Comment (hcchien) [inline fields/types/gcsimage/GcsImageType.js]: sort the keys?

File: fields/types/gcsimage/GcsImageType.js [hunk 11]
```diff
@@ -0,0 +1,361 @@
...
+            return gcsHelper.makeFilePublicPrivateRead(uploadedFile, isPublicRead);
+        }).then(function(response) {
+            var fileData = {
+                bucket: field.options.bucket,
+                filename: filename,
+                originalname: originalname,
+                path: path,
+                size: file.size,
+                filetype: filetype,
+                url: gcsHelper.getPublicUrl(field.options.bucket, path + filename),
```
Comment (hcchien) [inline fields/types/gcsimage/GcsImageType.js]: sort the keys

File: fields/types/gcsimage/GcsImageType.js [hunk 12]
```diff
@@ -0,0 +1,361 @@
...
+gcsimage.prototype.getRequestHandler = function(item, req, paths, callback) {
+
+    var field = this;
+
+    if (utils.isFunction(paths)) {
+        callback = paths;
+        paths = field.paths;
+    } else if (!paths) {
+        paths = field.paths;
+    }
```
Comment (hcchien) [inline fields/types/gcsimage/GcsImageType.js]: so the paths may be null or undefined?

File: fields/types/gcsimages/GcsImagesType.js [hunk 13]
```diff
@@ -0,0 +1,417 @@
...
+var path = require('path');
+var _ = require('underscore');
+var moment = require('moment');
+var keystone = require('../../../');
+var util = require('util');
+var utils = require('keystone-utils');
+var super_ = require('../Type');
+var async = require('async');
+var grappling = require('grappling-hook');
+var gcsHelper = require('../../../lib/gcsHelper');
```
Comment (hcchien) [inline fields/types/gcsimages/GcsImagesType.js]: variables sorting

File: fields/types/gcsimages/GcsImagesType.js [hunk 14]
```diff
@@ -0,0 +1,417 @@
...
+        originalname: this._path.append('.originalname'),
+        path: this._path.append('.path'),
+        size: this._path.append('.size'),
+        filetype: this._path.append('.filetype'),
+        url: this._path.append('.url'),
+        // virtuals
+        exists: this._path.append('.exists'),
+        upload: this._path.append('_upload'),
+        action: this._path.append('_action'),
+        order: this._path.append('_order'),
```
Comment (hcchien) [inline fields/types/gcsimages/GcsImagesType.js]: keys sorting

File: fields/types/gcsimages/GcsImagesType.js [hunk 15]
```diff
@@ -0,0 +1,417 @@
...
+
+    var schemaPaths = new mongoose.Schema({
+        bucket: String,
+        filename: String,
+        originalname: String,
+        path: String,
+        size: Number,
+        filetype: String,
+        url: String
+    });
```
Comment (hcchien) [inline fields/types/gcsimages/GcsImagesType.js]: keys sorting

File: fields/types/gcsimage/GcsImageType.js (hunk 12 above)
Comment (nickhsine) [inline fields/types/gcsimage/GcsImageType.js]: it won't. it will be the ```field.paths```

**Pull request merged**